compiler/rustc_codegen_ssa/src/coverageinfo/map.rs

```diff
@@ -238,34 +242,74 @@ impl<'tcx> FunctionCoverage<'tcx> {
                     id_to_counter(&new_indexes, rhs).map(|rhs_counter| (lhs_counter, rhs_counter))
                 })
             {
-                debug_assert!(
-                    (lhs_counter.id as usize)
-                        < usize::max(self.counters.len(), self.expressions.len())
-                );
-                debug_assert!(
-                    (rhs_counter.id as usize)
-                        < usize::max(self.counters.len(), self.expressions.len())
-                );
-                // Both operands exist. `Expression` operands exist in `self.expressions` and have
-                // been assigned a `new_index`.
-                let mapped_expression_index =
-                    MappedExpressionIndex::from(counter_expressions.len());
-                let expression = CounterExpression::new(
-                    lhs_counter,
-                    match op {
-                        Op::Add => ExprKind::Add,
-                        Op::Subtract => ExprKind::Subtract,
-                    },
-                    rhs_counter,
-                );
-                debug!(
-                    "Adding expression {:?} = {:?}, region: {:?}",
-                    mapped_expression_index, expression, optional_region
-                );
-                counter_expressions.push(expression);
-                new_indexes[original_index] = Some(mapped_expression_index);
-                if let Some(region) = optional_region {
-                    expression_regions.push((Counter::expression(mapped_expression_index), region));
+                if !(lhs_counter.is_zero() && op.is_subtract()) {
+                    debug_assert!(
+                        lhs_counter.is_zero()
+                            || ((lhs_counter.zero_based_id() as usize)
```

Comment:
Pedantic nit: This can wrap for `usize = u16`, but unclear if the compiler built for a platform where this holds true is going be at all useful.

I would perhaps write this as:

```rust
usize::max(self.counters.len(), self.expressions.len()))
    .try_into::<u32>()
    .map(|v| lhs_counter.zero_based_id() < v)
    .unwrap_or(true)
```

Though, I guess, if `self.counters.len() >= u32::MAX` things are probably already going very wrong everywhere anyway.

Reply:
I think I get it, but this is a lot harder to understand what's going on, and why. I think if `usize = u16` *_and_* the `zero_based_id() > usize::MAX` then a lot of things are probably breaking/broken. 

Plus this is in a `debug_assert!()`, so unless we're going to write a test case that creates thousands of branches to catch this, we're unlikely to even execute this check (in production code, particularly when targeting a small machine like this).

I'm tempted to leave this as-is to save others the mental exercise, but I can add a comment.

@nagisa - You said it's a "nit", but do let me know if you strongly disagree.

Reply:
FYI, I just added a comment above each `as usize` so at least the concern is captured there.